### ISSUE TITLE
RFC: Re-layer Draft SWA into Committed Sidecar KV and Verify Scratch

### ISSUE THREAD

**Issue by zhangxiaolei123456** [sgl-project/sglang#35789]
### Checklist

- [ ] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

### Summary

This RFC proposes splitting draft Sliding Window Attention (SWA) KV in speculative decoding, such as DSV4 DSpark, into two explicitly modeled components:

```text
Draft SWA
├── committed sidecar KV: content-scoped and reusable
└── verify temporary:     step-scoped scratch, never entering cache
```

Today, draft SWA is implicitly managed through the draft worker KV pool and coupled to target KV pool sizing. Its committed and in-flight states are not represented separately.

This causes two problems:

1. Committed draft KV lacks explicit cache semantics.
2. Draft SWA memory is estimated as if draft layers had the same average structure as target layers.

The proposal introduces:

- A content-scoped draft SWA sidecar aligned with the target prefix cache.
- Step-scoped scratch for speculative verification.
- SWA-only draft memory accounting.
- Correct stage-count derivation.
- Radix cache, PD, and HiCache integration.
- A draft-only compatibility path for `unified_kv`.

This proposal does not change model numerics. It changes memory ownership, cache lifecycle, physical layout, and pool sizing.

### 2.1 Memory ownership should follow data semantics

Device memory in SGLang can be classified by ownership and lifetime:

```text
Model-scoped:
  model weights and persistent model buffers

Content-scoped:
  committed KV, Mamba checkpoints, compressed pages

Request-scoped:
  active recurrent state and request-local runtime state

Step-scoped:
  verification intermediates, logits, and temporary workspaces
```

The important distinction is:

```text
Running / intermediate / in-flight state
  -> request-scoped or step-scoped

Committed / reusable artifact
  -> content-scoped
```

Draft SWA currently mixes these two categories.

### 2.2 Semantics of committed draft SWA

Taking DSV4 DSpark as an example, canonical draft KV is derived from injected target hidden states:

```text
target hidden states -> canonical draft KV
```

For an already committed prefix:

```text
same prefix + same target/draft weights -> same canonical draft KV
```

Therefore, committed draft SWA is a reusable artifact:

```text
committed draft SWA = content-scoped sidecar KV
```

Its lifecycle should follow the committed target prefix rather than the lifetime of a request slot.

### 2.3 Semantics of verify temporary

During speculative verification, temporary draft KV may represent tokens that are later rejected.

Before acceptance is known, that data is not part of the committed sequence:

```text
in-flight draft KV != committed prefix artifact
```

Verify temporary should therefore be:

```text
step-scoped scratch
```

It must not become radix-visible, transferable, or reusable until the corresponding tokens have been accepted and canonical committed KV has been produced.

### 3.1 Committed draft SWA is not explicitly modeled

The problem is not the target KV cache. The problem is that draft SWA is implicitly folded into the existing draft worker pool and target budget logic.

There is no independent abstraction expressing:

```text
this draft KV belongs to committed prefix content
```

Consequently:

1. Cache ownership is unclear.
2. Committed and speculative data share lifecycle assumptions.
3. Prefix reuse is difficult to express.
4. PD and HiCache integration is indirect.
5. Pool sizing is coupled to target-layer averages.

### 3.2 Draft SWA memory accounting is inaccurate

The current DSV4 pool configurator approximately scales the target cost as follows:

```text
bytes_per_full_token *=
    (target_layers + draft_layers) / target_layers
```

This treats each draft layer as one target average layer.

However, DSV4 DSpark draft layers are SWA-only:

```text
draft layer does not include C4
draft layer does not include C128
draft layer does not include the C4 indexer
draft layer does not include C4/C128 compression state
```

Charging a draft layer as a target average layer therefore overestimates draft KV memory and reduces usable target KV capacity.

A more accurate semantic term is:

```text
draft_swa_committed_bytes_per_full_token =
    swa_full_tokens_ratio
    * draft_swa_bytes_per_token
    * draft_swa_stage_count
```

The physical `draft_swa_bytes_per_token` must come from the active backend layout:

```text
paged backend:
  packed/padded SWA page bytes

unified backend:
  unified row stride and storage dtype
```

Verify scratch is a separate fixed workspace term:

```text
draft_swa_scratch_bytes =
    max_concurrent_requests
    * verify_scratch_width
    * scratch_bytes_per_row
    * draft_swa_stage_count
```

This separates content capacity from fixed speculative workspace.

### 3.3 DSpark stage count is ambiguous

A DSpark bundle may expose:

```text
num_nextn_predict_layers = 1
dspark_target_layer_ids = [40, 41, 42]
```

Although the model has one logical next-token prediction layer, the runtime captures and injects multiple target stages.

The draft SWA sidecar stage count must therefore come from the actual capture/injection stages:

```text
draft_swa_stage_count = len(dspark_target_layer_ids)
```

It must not be inferred only from `num_nextn_predict_layers`.

### 3.4 Rejected data can outlive its semantic scope

A request-scoped ring may physically retain rejected draft rows until they are overwritten.

Physical retention alone is acceptable for scratch, but those rows must never be interpreted as:

- Committed KV.
- Radix-cache content.
- PD-transferable state.
- HiCache-offloadable state.

The current representation does not make this boundary explicit.

### 4.1 Different architectural dimensions

This RFC and `unified_kv` address different concerns.

`unified_kv` primarily defines an execution-oriented physical layout:

```text
request slot
  -> fixed SWA ring
  -> position modulo ring size
```

Its strengths include:

- Predictable addresses.
- Contiguous memory.
- Low metadata overhead.
- Efficient unified SWA/C4/C128 kernels.
- Request-slot-oriented hot-path execution.

The draft sidecar proposal defines ownership and cache lifecycle:

```text
committed prefix artifact -> content-scoped sidecar
in-flight verify state     -> step-scoped scratch
```

The distinction is:

```text
unified_kv:
  execution-oriented physical layout

draft SWA sidecar:
  content-oriented ownership and lifecycle
```

### 4.2 This RFC does not redesign target `unified_kv`

The proposal does not convert target unified SWA into a content-addressed layout.

The following target structures remain unchanged:

```text
target unified SWA ring
target C4/C128 layout
target compression state
target unified attention kernels
target request-slot addressing
```

Compatibility is implemented as a draft-only extension.

### 4.3 Draft-only unified layout

A compatible draft unified pool can use:

```text
[ committed draft SWA rows ][ verify scratch rows ]
```

The draft attention metadata combines two address sources:

```text
committed prefix:
  radix/content mapping -> committed sidecar rows

current speculative step:
  request/step mapping -> scratch rows
```

The buffer can remain contiguous and compatible with a unified attention interface while preserving separate semantic lifecycles.

### 4.4 Transfer semantics under `unified_kv`

Target and draft state may use different transfer semantics:

```text
target unified SWA:
  request-scoped SWA_RING state

draft committed sidecar:
  content-scoped SWA page state

draft verify scratch:
  never transferred
```

This preserves the original target unified design while making committed draft SWA independently reusable.

### 5.1 Overall structure

```text
Draft SWA
├── committed sidecar KV
│   ├── content-scoped
│   ├── radix-visible
│   ├── reusable
│   └── transferable
└── verify temporary
    ├── step-scoped
    ├── allocator-independent
    ├── radix-invisible
    └── non-transferable
```

This follows the same principle used by Mamba and compressed KV artifacts:

```text
runtime temporary data is not shared
only committed artifacts enter the cache
```

### 5.2 Committed sidecar KV

Committed sidecar KV represents canonical draft SWA KV belonging to an accepted prefix.

Its properties are:

1. Content-scoped.
2. Determined by prefix content and model weights.
3. Reusable on radix-cache hits.
4. Coupled to target prefix lifecycle.
5. Eligible for PD transfer.
6. Eligible for HiCache offload/loadback.
7. Evicted consistently with the corresponding target prefix.

It can be understood as:

```text
the draft-side sidecar of the target prefix cache
```

The sidecar may use a separate pool or an additional physical region, but it must have explicit coverage and lifecycle semantics.

### 5.3 Verify temporary

Verify temporary contains draft KV and attention inputs needed only by the active speculative step.

Its properties are:

1. Step-scoped.
2. Never radix-visible.
3. Excluded from HiCache.
4. Excluded from PD transfer.
5. Excluded from long-lived allocator state.
6. Reusable as workspace after the step.
7. Incapable of producing cache artifacts for rejected tokens.

Possible physical representations include:

```text
temporary tensor
workspace buffer
step-local scratch range
```

It must not be treated as a long-lived committed request cache.

### 5.4 Accept-path data flow

The numerically safe baseline is:

```text
draft proposal / verify
  -> generate step-local temporary KV
  -> run verify attention
  -> compute accept/reject

accepted
  -> derive canonical draft KV from target hidden states
  -> write canonical KV into committed sidecar
  -> mark sidecar coverage as committed

rejected
  -> discard temporary KV
```

The invariant is:

```text
only canonical, committed draft KV enters the sidecar
```

### 5.5 Promotion optimization

Promotion is valid only when verify temporary already contains the canonical target-hidden-derived KV representation:

```text
canonical verify scratch
  -> select accepted rows
  -> promote/copy accepted rows into sidecar
```

Proposal KV produced from draft mask/noise hidden states is not necessarily numerically equivalent to canonical committed KV.

Therefore:

```text
proposal scratch must not be promoted directly
```

The first implementation should prefer direct canonical materialization. Promotion can be introduced later after numerical equivalence is demonstrated.

### 5.6 Prefix-cache-hit behavior

When both target and draft sidecar coverage hit:

```text
target committed KV hit
draft committed sidecar KV hit
  -> reuse both directly
```

When target KV hits but draft sidecar coverage is missing:

```text
target committed KV hit
draft sidecar miss
  -> recompute/backfill draft sidecar
```

For SWA, the fallback should preserve the reusable prefix and recompute only the required trailing window where possible:

```text
reuse prefix before trailing SWA window
recompute trailing window
commit reconstructed sidecar coverage
```

This fallback should be observable and should not become the normal path.

### 5.7 Eviction and consistency

Sidecar coverage must never outlive its corresponding target prefix.

The required invariant is:

```text
valid draft sidecar coverage
  implies
valid corresponding target prefix coverage
```

Insert, split, eviction, backup, and loadback must update sidecar coverage consistently.

A stale sidecar must never be paired with a newly allocated target prefix using the same physical slots.

### 6.1 Paged backend

The paged layout can use:

```text
committed sidecar pages:
  indexed through content/SWA mapping

verify scratch pages:
  indexed through request slot and step offset
```

Only committed pages are exposed to radix cache, PD, and HiCache.

### 6.2 `unified_kv` backend

The target unified layout remains unchanged.

The draft unified pool adds:

```text
committed content-addressed rows
verify scratch rows
```

The draft attention index stream is assembled from:

```text
content indices for committed prefix
scratch indices for current verify step
```

No target C4/C128 or target ring redesign is required.

### 6.3 Backend-independent semantics

Physical layouts may differ, but all backends must satisfy:

```text
committed rows:
  content-scoped and transferable

scratch rows:
  step-scoped and non-transferable
```

Backend-specific row stride, padding, and dtype must be reflected in memory accounting.

### 7.1 Accounting fix

Replace target-average-layer scaling with explicit draft SWA terms:

```text
committed sidecar bytes per full token
fixed verify scratch bytes
```

This step changes pool sizing only.

### 7.2 Stage-count correctness

Derive `draft_swa_stage_count` from actual target capture/injection stages, such as `dspark_target_layer_ids`.

### 7.3 Sidecar abstraction

Introduce an explicit draft SWA sidecar component with:

- Content coverage.
- Target-prefix coupling.
- Insert, split, and eviction lifecycle.
- Backup and loadback lifecycle.
- Radix visibility.

### 7.4 Verify scratch separation

Route speculative proposal and verify KV through step-local scratch.

Scratch addresses must not be inserted into request-to-token mappings or radix nodes.

### 7.5 Accept path

Materialize canonical accepted draft KV directly into committed sidecar rows.

Rejected rows produce no sidecar coverage.

### 7.6 Prefix-hit and fallback path

Reuse sidecar KV on a complete hit.

On a target-hit/sidecar-miss case, recompute only the required trailing SWA region where possible.

### 7.7 PD and HiCache integration

Expose committed sidecar pages as transferable state.

Explicitly exclude verify scratch from all transfer descriptors.

### 7.8 `unified_kv` adapter

Add a draft-only unified sidecar/scratch layout and mixed index generation.

Do not modify the target unified ring or compressed-state kernels.

### 7.9 Rollout

Gate the new layout behind a server argument during rollout.

Unsupported backends must reject explicit enablement rather than silently falling back to incompatible semantics.

### 8.1 Numerical compatibility

The design should not change model numerics.

Canonical committed KV must continue to come from the target-hidden injection path unless a promotion path has proven numerical equivalence.

### 8.2 Cache consistency

Sidecar and target prefix eviction must remain coupled.

Incorrect coverage updates could pair stale sidecar KV with unrelated target content.

### 8.3 Fallback overhead

A frequent sidecar-miss fallback would erase reuse benefits.

The implementation should record sidecar hit rate and trailing-recompute frequency.

### 8.4 Accept-path overhead

Extra materialization or copies may offset memory benefits.

The baseline should use one canonical projection and direct committed scatter. Promotion should only be added when it reduces measured critical-path cost.

### 8.5 Backend complexity

Paged and unified backends use different physical layouts.

The abstraction must standardize semantics without forcing target unified kernels into the paged cache model.

### 8.6 Memory accounting drift

Backend row padding, dtype, and page stride must be included in physical sizing.

Semantic token counts alone are insufficient for exact allocation.

### 9.1 Correctness

Verify output parity and speculative accept-rate parity against main.

### 9.2 Cache-artifact invariant

Confirm that:

```text
rejected draft tokens produce no radix artifact
verify scratch is unreachable after the step
only accepted canonical KV gains sidecar coverage
```

### 9.3 Accounting

Compare calculated draft memory against allocated and measured memory.

Expected results:

- Lower draft over-reservation.
- Larger usable target KV pool.
- Fixed scratch overhead independent of cached token count.

### 9.4 Prefix reuse

Measure:

- Sidecar hit rate.
- Target-hit/sidecar-miss rate.
- Trailing-window recompute frequency.
- Amount of reused committed draft KV.

### 9.5 PD and HiCache

Verify that committed sidecar KV survives:

- Prefill/decode transfer.
- Offload.
- Loadback.
- Eviction and reallocation.

Verify that scratch rows are never transferred.

### 9.6 `unified_kv`

Validate that:

- Target unified behavior is unchanged.
- Draft committed rows use content indices.
- Verify rows use scratch indices.
- Mixed index streams produce numerical parity.
- Target `SWA_RING` and draft `SWA` transfer states coexist correctly.

### 9.7 Performance

Measure:

- Per-step scratch overhead.
- Canonical commit cost.
- Optional promotion cost.
- Attention metadata construction.
- Prefix-hit latency.
- End-to-end throughput and latency.

### 10.1 Keep draft SWA request-scoped

This preserves the current implementation shape but prevents committed draft KV from becoming a reusable prefix artifact.

### 10.2 Recompute draft SWA on every prefix hit

This avoids sidecar memory but makes fallback the default and repeatedly pays the target-hidden projection cost.

### 10.3 Keep verify data in a long-lived request ring

This simplifies addressing but retains rejected state longer than necessary and blurs committed versus in-flight ownership.

### 10.4 Convert target unified SWA to content-addressed storage

This would unnecessarily redesign the target unified hot path and is outside the scope of this RFC.

The preferred approach is a draft-only unified extension.

### 10.5 Promote proposal scratch directly

This is invalid unless proposal scratch is numerically identical to canonical target-hidden-derived KV.

The proposal does not assume this equivalence.
